### DISCUSSION TITLE
3D + periodic BC tracking?

### DISCUSSION THREAD

**Discussion by keli559** [Parcels-code/Parcels#1271] [Q&A]
Ran into a problem with 3D data (time depth lat lon) where depth lat lon are in meters, with periodic BCs at the south north and at the west east. For some reason, the pset as initially set up have 27 in total, 3 x 3 horizontally, and 3 vertical layers. but at the end, only got results from 3x 3 at z =0 m. The other 18 particles are gone. 

The background circulation model output setup is:
`fieldset = FieldSet.from_nemo(filenames\
                              , variables, dimensions, mesh = 'flat'\
                                            , allow_time_extrapolation=True)`
The particles are set up as:
`pset = ParticleSet.from_list(fieldset=fieldset, pclass=JITParticle\
                             , lon = xpi, lat = ypi, depth = zpi\
                             , time = np.zeros_like(xpi))`
When punching in 'pset', it shows :
`P[162](lon=0.000000, lat=0.000000, depth=-70.000000, time=0.000000)
P[163](lon=0.000000, lat=0.000000, depth=-35.000000, time=0.000000)
P[164](lon=0.000000, lat=0.000000, depth=0.000000, time=0.000000)
P[165](lon=0.000000, lat=64.000000, depth=-70.000000, time=0.000000)
P[166](lon=0.000000, lat=64.000000, depth=-35.000000, time=0.000000)
P[167](lon=0.000000, lat=64.000000, depth=0.000000, time=0.000000)
... ... ... ...`

pset.execute is:

`pset.execute(AdvectionRK4_3D + pset.Kernel(periodicBC)
             , runtime=delta(hours=dT/3600.0)\
             , dt=delta(hours=timeStep/3600.0)\
             , output_file=output_file\
                              , recovery={ErrorCode.ErrorOutOfBounds: DeleteParticle})`

where periodicBC function was borrowed from the tutorial of Oceanparcels. 
`It does give me a result, but only z=0
P[164](lon=34.504272, lat=123.591668, depth=0.000000, time=299700.000000)
P[167](lon=122.359932, lat=118.575767, depth=0.000000, time=299700.000000)
P[170](lon=34.504272, lat=123.591668, depth=0.000000, time=299700.000000)
P[173](lon=109.732106, lat=123.345966, depth=0.000000, time=299700.000000)
P[176](lon=62.235887, lat=9.392261, depth=0.000000, time=299700.000000)
P[179](lon=109.732106, lat=123.345966, depth=0.000000, time=299700.000000) ... ... ... ...`
The other 18 particles inside underwater are gone. ...

**Comment by erikvansebille** [comment]
Hi @keli559, thanks for using Parcels! Sorry, but with this info I can't figure out what's happening. What is the error you get when you don't use the `recovery={ErrorCode.ErrorOutOfBounds: DeleteParticle}` part? What happens if you remove the periodicBC Kernel? And can you give more information about the fieldset?
